criteria.c

```diff
@@ -0,0 +1,317 @@
+#define _POSIX_C_SOURCE 200809L
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <strings.h>
+#include <wayland-client.h>
+#include "mako.h"
+#include "config.h"
+#include "criteria.h"
+#include "notification.h"
+
+struct mako_criteria *create_criteria(struct mako_config *config) {
+	struct mako_criteria *criteria = calloc(1, sizeof(struct mako_criteria));
+	if (criteria == NULL) {
+		fprintf(stderr, "allocation failed\n");
+		return NULL;
+	}
+
+	wl_list_insert(config->criteria.prev, &criteria->link);
+	return criteria;
+}
+
+void destroy_criteria(struct mako_criteria *criteria) {
+	wl_list_remove(&criteria->link);
+
+	finish_style(&criteria->style);
+	free(criteria->app_name);
+	free(criteria->app_icon);
+	free(criteria->category);
+	free(criteria->desktop_entry);
+	free(criteria);
+}
+
+bool match_criteria(struct mako_criteria *criteria,
+		struct mako_notification *notif) {
+	struct mako_criteria_spec spec = criteria->spec;
+
+	if (spec.app_name &&
+			strcmp(criteria->app_name, notif->app_name) != 0) {
+		return false;
+	}
+
+	if (spec.app_icon &&
+			strcmp(criteria->app_icon, notif->app_icon) != 0) {
+		return false;
+	}
+
+	if (spec.actionable &&
+			criteria->actionable == wl_list_empty(&notif->actions)) {
+		return false;
+	}
+
+	if (spec.urgency &&
+			criteria->urgency != notif->urgency) {
+		return false;
+	}
+
+	if (spec.category &&
+			strcmp(criteria->category, notif->category) != 0) {
+		return false;
+	}
+
+	if (spec.desktop_entry &&
+			strcmp(criteria->desktop_entry, notif->desktop_entry) != 0) {
+		return false;
+	}
+
+	return true;
+}
+
+bool parse_boolean(const char *string, bool *out) {
+	if (strcasecmp(string, "true") == 0 || strcmp(string, "1") == 0) {
+		*out = true;
+		return true;
+	}
+	else if (strcasecmp(string, "false") == 0 || strcmp(string, "0") == 0) {
+		*out = false;
+		return true;
+	}
+
+	return false;
+}
+
+bool parse_urgency(const char *string, enum mako_notification_urgency *out) {
+	if (strcasecmp(string, "low") == 0) {
+		*out = MAKO_NOTIFICATION_URGENCY_LOW;
+		return true;
+	}
+	else if (strcasecmp(string, "normal") == 0) {
+		*out = MAKO_NOTIFICATION_URGENCY_NORMAL;
+		return true;
+	}
+	else if (strcasecmp(string, "high") == 0) {
+		*out = MAKO_NOTIFICATION_URGENCY_HIGH;
+		return true;
+	}
+
+	return false;
+}
+
+bool parse_criteria(const char *string, struct mako_criteria *criteria) {
+	// Create space to build up the current token that we're reading. We know
+	// that no single token can ever exceed the length of the entire criteria
+	// string, so that's a safe length to use for the buffer.
+	int token_max_length = strlen(string) + 1;
+	char token[token_max_length];
+	memset(token, 0, token_max_length);
```

Comment:
Instead of calling `memset`, you can initialize `token` with `{0}` and all fields will be set to zero.

Reply:
You can't use initialization syntax with a variable-length array. :(

Reply:
Oh, I didn't know that. RIP